airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/NamespacingMapper.java

```diff
@@ -41,39 +43,59 @@
+
   private final NamespaceDefinitionType namespaceDefinition;
-  private final String prefix;
+  private final String namespaceFormat;
+  private final String streamPrefix;
 
-  public NamespacingMapper(NamespaceDefinitionType namespaceDefinition, String prefix) {
+  public NamespacingMapper(NamespaceDefinitionType namespaceDefinition, String namespaceFormat, String streamPrefix) {
     this.namespaceDefinition = namespaceDefinition;
-    this.prefix = prefix;
+    this.namespaceFormat = namespaceFormat;
+    this.streamPrefix = streamPrefix;
   }
 
   @Override
   public ConfiguredAirbyteCatalog mapCatalog(final ConfiguredAirbyteCatalog inputCatalog) {
     final ConfiguredAirbyteCatalog catalog = Jsons.clone(inputCatalog);
-    catalog.getStreams().forEach(s -> mutateStream(s.getStream(), namespaceDefinition, prefix));
+    catalog.getStreams().forEach(s -> {
+      final AirbyteStream stream = s.getStream();
+      if (namespaceDefinition != null) {
+        if (namespaceDefinition.equals(NamespaceDefinitionType.DESTINATION)) {
+          stream.withNamespace(null);
+        } else if (namespaceDefinition.equals(NamespaceDefinitionType.CUSTOMFORMAT)) {
+          stream.withNamespace(formatNamespace(stream.getNamespace(), namespaceFormat));
+        }
+      }
+      stream.withName(transformStreamName(stream.getName(), streamPrefix));
+    });
     return catalog;
   }
 
   @Override
   public AirbyteMessage mapMessage(final AirbyteMessage inputMessage) {
     if (inputMessage.getType() == Type.RECORD) {
       final AirbyteMessage message = Jsons.clone(inputMessage);
-      if (namespaceDefinition != null && namespaceDefinition.equals(NamespaceDefinitionType.DESTINATION)) {
-        message.getRecord().withNamespace(null);
+      if (namespaceDefinition != null) {
+        if (namespaceDefinition.equals(NamespaceDefinitionType.DESTINATION)) {
+          message.getRecord().withNamespace(null);
+        } else if (namespaceDefinition.equals(NamespaceDefinitionType.CUSTOMFORMAT)) {
+          message.getRecord().withNamespace(formatNamespace(message.getRecord().getNamespace(), namespaceFormat));
+        }
       }
-      message.getRecord().setStream(transformStreamName(message.getRecord().getStream(), prefix));
+      message.getRecord().setStream(transformStreamName(message.getRecord().getStream(), streamPrefix));
       return message;
     }
     return inputMessage;
   }
 
-  private static void mutateStream(final AirbyteStream stream, final NamespaceDefinitionType namespaceDefinition, final String prefix) {
-    if (namespaceDefinition != null && namespaceDefinition.equals(NamespaceDefinitionType.DESTINATION)) {
-      stream.withNamespace(null);
+  private static String formatNamespace(final String sourceNamespace, final String namespaceFormat) {
+    if (Strings.isNotEmpty(namespaceFormat)) {
+      return namespaceFormat.replaceAll("%s", Strings.isNotEmpty(sourceNamespace) ? sourceNamespace : "");
+    } else {
```

Comment:
suggestion, move this earlier to avoid the else.

Reply:
e.g.
```
if (Strings.isEmpty(namespaceFormat) {
      LOGGER.error("Namespace Format cannot be empty. Falling back to default namespace from destination settings");
      return null;
}

return namespaceFormat.replaceAll();
```

Reply:
i'll put to the end so if the replace results into an empty string, the error will be logged